File: app/src/main/java/com/haishinkit/rtmp/messages/RTMPAcknowledgementMessage.java
package com.haishinkit.rtmp.messages;

import com.haishinkit.rtmp.RTMPConnection;
import com.haishinkit.rtmp.RTMPSocket;
import java.nio.ByteBuffer;

/**
 * 5.4.3 Acknowledgement (3)
 */
public final class RTMPAcknowledgementMessage extends RTMPMessage {
    private static final int CAPACITY = 4;

    private byte[] sequence = null;

    public RTMPAcknowledgementMessage() {
        super(Type.ACK);
        sequence = new byte[CAPACITY];
    }

    public final byte[] getSequence() {
        return sequence;
    }

    public final RTMPAcknowledgementMessage setSequence(final byte[] sequence) {
        this.sequence = sequence;
        return this;
    }

    @Override
    public ByteBuffer encode(final RTMPSocket socket)  {
        if (socket == null) {
            throw new IllegalArgumentException();
        }
        ByteBuffer buffer = ByteBuffer.allocate(CAPACITY);
        buffer.put(sequence);
        return buffer;
    }

    @Override
    public RTMPMessage decode(final ByteBuffer buffer) {
        if (buffer == null) {
            throw new IllegalArgumentException();
        }
        buffer.get(getSequence());
        return this;
    }

    @Override
    public RTMPMessage execute(final RTMPConnection connection) {
        if (connection == null) {
            throw new IllegalArgumentException();
        }
        return this;
    }
}

File: haishinkit/src/main/java/com/haishinkit/rtmp/messages/RTMPAcknowledgementMessage.kt
package com.haishinkit.rtmp.messages

import com.haishinkit.rtmp.RTMPConnection
import com.haishinkit.rtmp.RTMPSocket
import java.nio.ByteBuffer

/**
 * 5.4.3 Acknowledgement (3)
 */
internal class RTMPAcknowledgementMessage : RTMPMessage(RTMPMessage.Type.ACK) {
    var sequence: ByteArray? = null
        private set

    init {
        sequence = ByteArray(CAPACITY)
    }

    override fun encode(socket: RTMPSocket): ByteBuffer {
        val buffer = ByteBuffer.allocate(CAPACITY)
        buffer.put(sequence)
        return buffer
    }

    override fun decode(buffer: ByteBuffer): RTMPMessage {
        buffer.get(sequence)
        return this
    }

    override fun execute(connection: RTMPConnection): RTMPMessage {
        return this
    }

    companion object {
        private val CAPACITY = 4
    }
}
